From adbc0c1530ff1111a4314b815f2620b0a03e7770 Mon Sep 17 00:00:00 2001
From: Aditya Kumar Singh <quic_adisi@quicinc.com>
Date: Tue, 31 Oct 2023 14:21:04 +0530
Subject: [PATCH 1/4] hostapd: fix deferred first link BSS's radius, auth
 server and 802.1X initialization

Currently, radius client, auth server and 802.1X are copied from first link's
BSS into the non-first link during its setup. However, there could be a case
where the first link is not initialized fully (ACS/HT40 SCAN/DFS). Hence, in such
cases, NULL is getting copied and later it leads to segmentation fault.

Add changes to initialize those on behalf of the first link if such be the case
and update it so that next time other non-first link can use it. If the interface
is not the first BSS, then also make the link which is initializing these as the
first link of that interface.

Signed-off-by: Aditya Kumar Singh <quic_adisi@quicinc.com>
---
 src/ap/authsrv.c    |  32 +++++++++++--
 src/ap/authsrv.h    |   2 +-
 src/ap/hostapd.c    | 108 +++++++++++++++++++++++++++++++-------------
 src/ap/ieee802_1x.c |  25 +++++++++-
 src/ap/ieee802_1x.h |   2 +-
 5 files changed, 128 insertions(+), 41 deletions(-)

--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -1345,6 +1345,57 @@ static int hostapd_start_beacon(struct h
 	return 0;
 }
 
+#ifndef CONFIG_NO_RADIUS
+static int hostapd_bss_radius_init(struct hostapd_data *hapd)
+{
+	struct hostapd_bss_config *conf;
+
+	if (!hapd)
+		return -1;
+
+	conf = hapd->conf;
+
+	if (hapd->radius) {
+		wpa_printf(MSG_DEBUG, "Skipping RADIUS client init (already done)");
+		return 0;
+	}
+
+	hapd->radius = radius_client_init(hapd, conf->radius);
+	if (!hapd->radius) {
+		wpa_printf(MSG_ERROR,
+			   "RADIUS client initialization failed.");
+		return -1;
+	}
+
+	if (conf->radius_das_port) {
+		struct radius_das_conf das_conf;
+
+		os_memset(&das_conf, 0, sizeof(das_conf));
+		das_conf.port = conf->radius_das_port;
+		das_conf.nas_identifier = conf->nas_identifier;
+		das_conf.shared_secret = conf->radius_das_shared_secret;
+		das_conf.shared_secret_len =
+			conf->radius_das_shared_secret_len;
+		das_conf.client_addr = &conf->radius_das_client_addr;
+		das_conf.time_window = conf->radius_das_time_window;
+		das_conf.require_event_timestamp =
+			conf->radius_das_require_event_timestamp;
+		das_conf.require_message_authenticator =
+			conf->radius_das_require_message_authenticator;
+		das_conf.ctx = hapd;
+		das_conf.disconnect = hostapd_das_disconnect;
+		das_conf.coa = hostapd_das_coa;
+		hapd->radius_das = radius_das_init(&das_conf);
+		if (!hapd->radius_das) {
+			wpa_printf(MSG_ERROR,
+				   "RADIUS DAS initialization failed.");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+#endif /* CONFIG_NO_RADIUS */
 
 /**
  * hostapd_setup_bss - Per-BSS setup (initialization)
@@ -1582,47 +1633,25 @@ setup_mld:
 #endif /* CONFIG_SQLITE */
 
 	if (hostapd_mld_is_first_bss(hapd)) {
-		hapd->radius = radius_client_init(hapd, conf->radius);
-		if (!hapd->radius) {
-			wpa_printf(MSG_ERROR,
-				   "RADIUS client initialization failed.");
+		if (hostapd_bss_radius_init(hapd))
+			return -1;
+	} else {
+#ifdef CONFIG_IEEE80211BE
+		struct hostapd_data *f_bss = hostapd_mld_get_first_bss(hapd);
+		if (!f_bss)
 			return -1;
-		}
 
-		if (conf->radius_das_port) {
-			struct radius_das_conf das_conf;
+		if (!f_bss->radius) {
+			wpa_printf(MSG_DEBUG,
+				   "MLD: First BSS RADIUS client does not exist. Init on its behalf");
 
-			os_memset(&das_conf, 0, sizeof(das_conf));
-			das_conf.port = conf->radius_das_port;
-			das_conf.nas_identifier = conf->nas_identifier;
-			das_conf.shared_secret = conf->radius_das_shared_secret;
-			das_conf.shared_secret_len =
-				conf->radius_das_shared_secret_len;
-			das_conf.client_addr = &conf->radius_das_client_addr;
-			das_conf.time_window = conf->radius_das_time_window;
-			das_conf.require_event_timestamp =
-				conf->radius_das_require_event_timestamp;
-			das_conf.require_message_authenticator =
-				conf->radius_das_require_message_authenticator;
-			das_conf.ctx = hapd;
-			das_conf.disconnect = hostapd_das_disconnect;
-			das_conf.coa = hostapd_das_coa;
-			hapd->radius_das = radius_das_init(&das_conf);
-			if (!hapd->radius_das) {
-				wpa_printf(MSG_ERROR,
-					   "RADIUS DAS initialization failed.");
-				return -1;
-			}
+			if (hostapd_bss_radius_init(f_bss))
+ 				return -1;
 		}
-	} else {
-#ifdef CONFIG_IEEE80211BE
-		struct hostapd_data *f_bss;
 
 		wpa_printf(MSG_DEBUG,
 			   "MLD: Using RADIUS client of the first BSS");
-		f_bss = hostapd_mld_get_first_bss(hapd);
-		if (!f_bss)
-			return -1;
+
 		hapd->radius = f_bss->radius;
 		hapd->radius_das = f_bss->radius_das;
 #endif /* CONFIG_IEEE80211BE */
@@ -2074,7 +2103,6 @@ static int setup_interface(struct hostap
 	 */
 	for (i = 1; i < iface->num_bss; i++) {
 		iface->bss[i]->driver = hapd->driver;
-		iface->bss[i]->drv_priv = hapd->drv_priv;
 	}
 
 	if (hostapd_validate_bssid_configuration(iface))
--- a/src/ap/ieee802_1x.c
+++ b/src/ap/ieee802_1x.c
@@ -2545,14 +2545,21 @@ int ieee802_1x_init(struct hostapd_data
 
 #ifdef CONFIG_IEEE80211BE
 	if (!hostapd_mld_is_first_bss(hapd)) {
-		struct hostapd_data *first;
+		struct hostapd_data *first = hostapd_mld_get_first_bss(hapd);
+		if (!first)
+			return -1;
+
+		if (!first->eapol_auth) {
+			wpa_printf(MSG_DEBUG,
+				  "MLD: First BSS IEEE 802.1X state machine does not exist. Init on its behalf");
+
+			if (ieee802_1x_init(first))
+				return -1;
+		}
 
 		wpa_printf(MSG_DEBUG,
-			   "MLD: Using IEEE 802.1X state machine of the first BSS");
+				"MLD: Using IEEE 802.1X state machine of the first BSS");
 
-		first = hostapd_mld_get_first_bss(hapd);
-		if (!first)
-			return -1;
 		hapd->eapol_auth = first->eapol_auth;
 		return 0;
 	}
--- a/src/ap/authsrv.c
+++ b/src/ap/authsrv.c
@@ -258,13 +258,21 @@ int authsrv_init(struct hostapd_data *ha
 {
 #ifdef CONFIG_IEEE80211BE
 	if (!hostapd_mld_is_first_bss(hapd)) {
-		struct hostapd_data *first;
+		struct hostapd_data *first = hostapd_mld_get_first_bss(hapd);
 
-		wpa_printf(MSG_DEBUG, "MLD: Using auth_serv of the first BSS");
-
-		first = hostapd_mld_get_first_bss(hapd);
 		if (!first)
 			return -1;
+
+		if (!first->eap_cfg) {
+			wpa_printf(MSG_DEBUG,
+				   "MLD: First BSS auth_serv does not exist. Init on its behalf");
+
+			if (authsrv_init(first))
+				return -1;
+		}
+
+		wpa_printf(MSG_DEBUG, "MLD: Using auth_serv of the first BSS");
+
 #ifdef EAP_TLS_FUNCS
 		hapd->ssl_ctx = first->ssl_ctx;
 #endif /* EAP_TLS_FUNCS */
